### PR TITLE
Feature/literal parametrized dtype

### PR DESCRIPTION
**Pull request by FlorianDeconinck**
## Description

In order to run a 32-bit float version of the model, we need at minima to be able to control literal dtype. This PR implements option 4 of (the gtscript concept workshop)[https://github.com/GridTools/concepts/blob/master/collaboration/gtscript-workshop/GTScript-Syntax-Discussion.md#gtscript-syntax-discussed-issues-20200829]

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/main/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich

### CODE REVIEW

**Review by havogt — CHANGES_REQUESTED** [review]
Just blocking the merge, until we have a new CAA.

**Comment by gronerl** [comment]
bors try

**Comment by gronerl** [comment]
bors try

File: src/gt4py/cartesian/frontend/defir_to_gtir.py [hunk 1]
```diff
@@ -82,9 +84,12 @@ def _make_literal(v: numbers.Number) -> gtir.Literal:
             dtype = common.DataType.INT64
         elif isinstance(v, float):
             dtype = common.DataType.FLOAT64
+        elif isinstance(v, np.float32):
+            dtype = common.DataType.FLOAT32
         else:
             print(
-                f"Warning: Only INT64 and FLOAT64 literals are supported currently. Implicitly upcasting `{v}` to FLOAT64"
+                "Warning: Only INT64, FLOAT64 and FLOAT32 (via np.float32) literals are "
+                f"supported currently. Implicitly upcasting `{v}` of type {type(v)} to FLOAT64"
```
Comment (gronerl) [inline src/gt4py/cartesian/frontend/defir_to_gtir.py:92]: Why not support all the integer types as well?

File: src/gt4py/cartesian/frontend/gtscript_frontend.py [hunk 2]
```diff
@@ -999,7 +1001,12 @@ def visit_Constant(
                 loc=nodes.Location.from_ast_node(node),
             )
         elif isinstance(value, numbers.Number):
-            data_type = nodes.DataType.from_dtype(np.dtype(type(value)))
+            value_type = (
+                self.dtypes[type(value)]
+                if self.dtypes and type(value) in self.dtypes.keys()
+                else np.dtype(type(value))
+            )
```
Comment (gronerl) [inline src/gt4py/cartesian/frontend/gtscript_frontend.py:1008]: There should be a check that raises if keys other than `int` or `float` are used.

**Review by gronerl — CHANGES_REQUESTED** [review]
Thanks for the PR. 

Regarding the testing of defir_to_gtir: If it is too much effort, I won't block because of this. After all, it is also not tested in the current form.

File: src/gt4py/cartesian/gtscript.py [hunk 3]
```diff
@@ -158,8 +158,8 @@ def stencil(
     definition=None,
     *,
     build_info=None,
-    dtypes=None,
     externals=None,
+    dtypes=None,
```
[the remark this answers is not in the document]
Comment (FlorianDeconinck) [reply src/gt4py/cartesian/gtscript.py:162]: Felt like external is a more "prevalent" parameter than dtypes. Debatable change, will revert

File: src/gt4py/cartesian/frontend/gtscript_frontend.py (hunk 2 above)
Comment (FlorianDeconinck) [reply src/gt4py/cartesian/frontend/gtscript_frontend.py:1008]: raise GTScriptSyntaxError

File: src/gt4py/cartesian/frontend/defir_to_gtir.py (hunk 1 above)
Comment (FlorianDeconinck) [reply src/gt4py/cartesian/frontend/defir_to_gtir.py:92]: Actually this was a quiet upcast, which is not great. So I added ints and turned unknown type into a TypeError, wadya think?

File: src/gt4py/cartesian/frontend/gtscript_frontend.py (hunk 2 above)
Comment (FlorianDeconinck) [reply src/gt4py/cartesian/frontend/gtscript_frontend.py:1008]: Actually no, you can do `FloatField["my_string"]` then `dtypes={"my_string":float}`, so not a raise

File: tests/cartesian_tests/unit_tests/test_gtc_backend/test_defir_to_gtir.py [hunk 4]
```diff
@@ -170,3 +171,22 @@ def test_field_decl_dims(defir_to_gtir, axes, expected_mask):
     field_decl = FieldDecl(name="a", data_type=DataType.INT64, axes=axes, is_api=True)
     gtir_decl = defir_to_gtir.visit_FieldDecl(field_decl)
     assert gtir_decl.dimensions == expected_mask
+
+
+def test_make_literal(defir_to_gtir):
```
Comment (gronerl) [inline tests/cartesian_tests/unit_tests/test_gtc_backend/test_defir_to_gtir.py:176]: What's up with the argument here?

Comment (FlorianDeconinck) [reply tests/cartesian_tests/unit_tests/test_gtc_backend/test_defir_to_gtir.py:176]: Yeah this can die

**Comment by gronerl** [comment]
bors try

**Comment by gronerl** [comment]
Ready to merge once CAA is cleared.

**Comment by FlorianDeconinck** [comment]
Cheers @gronerl!

**Review by havogt — APPROVED** [review]
Thanks for your patience! Let's add your new affiliation to the authors list.

**Comment by gronerl** [comment]
@FlorianDeconinck can you add your name to https://github.com/FlorianDeconinck/gt4py/blame/feature/literal_parametrized_dtype/AUTHORS.rst as appropriate?

**Comment by gronerl** [comment]
We were asked to hold back with merging until #1150 is merged. We will still get this in for the next version tag.

**Pull request merged**